File: project-euler.java/src/main/java/euler/solutions/_23.java
package euler.solutions;

import euler.common.BaseSolution;
import euler.common.Result;
import gnu.trove.iterator.TIntIterator;
import gnu.trove.list.TIntList;
import gnu.trove.list.array.TIntArrayList;

import static euler.common.NumberUtils.getDivisors;
import static euler.common.NumberUtils.sum;

public class _23 extends BaseSolution {
    TIntList  abundants;
    boolean[] isAbundants;

    @Override
    public Result solve() {
        findAbundants();

        long sum = 0;
        for (int n = 1; n <= 28123; n++) {
            if (!isAbundantSum(n)) {
                sum += n;
            }
        }

        return new Result(sum);
    }

    private void findAbundants() {
        abundants =  new TIntArrayList(10000);
        isAbundants = new boolean[28123];

        for (int num = 2; num <= 28123; num++) {
            if (sum(getDivisors(num)) > num) {
                abundants.add(num);
                isAbundants[num - 1] = true;
            }
        }
    }

    private boolean isAbundantSum(int n) {
        for (TIntIterator it = abundants.iterator(); it.hasNext();) {
            int abundant = it.next();
            if (abundant >= n) {
                return false;
            }
            if (isAbundants[n - abundant -1]) {
                return true;
            }
        }
        return false;
    }
}

File: project-euler.go/solutions/_23/23.go
package _23

import "project-euler/common/numbers"

var abundants   []int
var isAbundants []bool

func Solve() int64 {
	findAbundants()

	var sum int64
	for n := 1; n <= 28123; n++ {
		if !isAbundantSum(n) {
			sum += int64(n)
		}
	}

	return sum
}

func findAbundants() {
	abundants   = make([]int,  0, 10000)
	isAbundants = make([]bool, 28123)

	for num := 2; num <= 28123; num++ {
		if numbers.Sum(numbers.GetDivisors(num)) > num {
			abundants = append(abundants, num)
			isAbundants[num - 1] = true
		}
	}
}

func isAbundantSum(n int) bool {
	for _,abundant := range abundants {
		if (abundant >= n) {
			return false
		}
		if (isAbundants[n - abundant - 1]) {
			return true
		}
	}
	return false
}
